### ISSUE TITLE
’EXISTS' cannot return accuracy results, influenced by the pattern order insider.

### ISSUE THREAD

**Issue by AndrewChung-GitHub** [nilp0inter/experta#14]
```py
from experta import *


class Goal(Fact):
    pass


class Hero(Fact):
    name = Field(str)
    status = Field(str, default="unoccupied")


class KE(KnowledgeEngine):
    @DefFacts()
    def goal_and_heroes(self):
        yield Goal('save-the-day')
        yield Hero(name="Death Defying Man")
        yield Hero(name="Stupendous Man")
        yield Hero(name="Incredible Man")

    @Rule(
        Goal('save-the-day'),
        EXISTS(            # bugs: return results influenced by the patterns order
            Hero(name="Death Defying Man"),
            Hero(status='occupied'),
            # Hero(status='occupied'),
            # Hero(name="Death Defying Man")
        )
    )
    def save_the_day(self):
        print("The day is saved")


if __name__ == '__main__':
    ke = KE()
    ke.reset()
    watch('RULES')
    ke.run()
```
Result:
```
The day is saved
INFO:experta.watchers.RULES:FIRE 1 save_the_day: <f-1>, <f-0>
```
However, if I reverse the patterns order in`EXISTS`, the result would be empty.
```
```

**Comment by nilp0inter** [comment]
Hi,

I am trying to reproduce the issue but in my tests the rule:

```python
    @Rule(
        Goal('save-the-day'),
        EXISTS(
            Hero(name="Death Defying Man"),
            Hero(status='unoccupied'),
        )
    )
    def save_the_day(self):
        print("The day is saved")
```

and

```python
    @Rule(
        Goal('save-the-day'),
        EXISTS(
            Hero(status='unoccupied'),
            Hero(name="Death Defying Man"),
        )
    )
    def save_the_day(self):
        print("The day is saved")
```

both yields the same result.

Which ``experta`` version are you using?

**Comment by AndrewChung-GitHub** [comment]
```
experta.__version__
```
'1.9.4'

please take note I used `Hero(status='occupied')` nor 'unoccupied'. Due to `Hero(status='occupied')` hadn't be defined above, it didn't exist. Could you try again, thanks.

**Comment by nilp0inter** [comment]
Oh, I think I understand. The example is not very clear, but the behavior is correct.

Let me explain: In the definition of ``Hero`` the field ``status`` has a default value ``unoccupied``. That means that any declared ``Hero`` fact that lacks the ``status`` field will have an implicit value of ``unoccupied``.

This is why the rule in the original example matched and yours don't. The explicit version would be:

```python
    @DefFacts()
    def goal_and_heroes(self):
        yield Goal('save-the-day')
        yield Hero(name="Death Defying Man", status="unoccupied")
        yield Hero(name="Stupendous Man", status="unoccupied")
        yield Hero(name="Incredible Man", status="unoccupied")
```

In your case the rule is not matching because **there are no** heroes with ``status="occupied"``.

**Comment by nilp0inter** [comment]
I modified[ the example](https://github.com/nilp0inter/experta/blob/develop/docs/examples/EXISTS.ipynb) to remove the implicit value. Please let me know if this way is more clear.

**Comment by AndrewChung-GitHub** [comment]
Thanks for your quickly response, however, you may misunderstood my point.
My concern is on the function of 'EXIST', I think as long as there is one or more facts matching, it should continue to execute following actions. However, in this case, it yields different results.

**Comment by nilp0inter** [comment]
You are very welcome, thank you for using *experta*.

For EXISTS to match, there must be **at least one instance of each fact** inside the EXISTS clause. As you can see the rule is not satisfied because none of the declared facts has ``status="occupied"`` in it.

**Comment by AndrewChung-GitHub** [comment]
yeah, indeed the facts `status="occupied"` didn't exist, but the fact `Hero(name="Death Defying Man")` always exists. Hence, I think no matter the order of these two patterns adjusted, the results should be the same and output the result of `print("The day is saved")`.

**Comment by nilp0inter** [comment]
The behavior that you are describing is not the behavior of the EXISTS clause but the OR clause. Please try the OR clause to see if it meets your requirements.

Also, check #8 for another clause that can be helpful to you.

**Comment by AndrewChung-GitHub** [comment]
Thanks for your patient explanations.

Yeah, I have not yet fully understood the function of "EXISTS", and fall into confusion about the descriptions in your reference file. 

> This CE receives a pattern and matches if one or more facts matches this pattern. This will match only once while one or more matching facts exists and will stop matching when there is no matching facts.

Could you give me more explanations and introductions on application scenarios?
Thanks again.
